### ISSUE TITLE
Pictures are not saved in an easy accessible place

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-android#1739]
If I "Save a media" (a picture here), it seems to put it in a place which is not easily available. I expect to have it in "Downloads" or "Pictures" like if I would take a picture with the camera or download a file with the browser. I have an Android 10.

**Comment by GearGit** [comment]
@kelson42 I would like to work on this issue

**Comment by Frans-Lukas** [comment]
@GearGit Saving directly to a public external media directory (ex. Pictures or Downloads) requires additional WRITE_EXTERNAL_STORAGE permissions which might not be wanted. What it is doing now is saving the media to the applications external media storage (does not require permissions) which will be deleted if the app is uninstalled. [See this link ](https://developer.android.com/training/data-storage#top_of_page) for external storage and permission table. Is this feature still wanted?

**Comment by GearGit** [comment]
@Frans-Lukas So Should i work on this issue or not?

**Comment by kelson42** [comment]
Waiting to @macgills to discuss the tech. Solution.

**Comment by macgills** [comment]
@GearGit was first to request so I am happy to give it to them. This is very much an Android 10 problem I believe so it will require a 10 device or running the emulator to solve I would think.

I could not reproduce the issue @kelson42 was experiencing, using our existing code a picture saved for me and was instantly available in `Google Photos` under my albums, though it was named poorly.

The existing code can be found at `KiwixWebView.SaveHandler`

**Comment by GearGit** [comment]
> @GearGit was first to request so I am happy to give it to them. This is very much an Android 10 problem I believe so it will require a 10 device or running the emulator to solve I would think.
> 
> I could not reproduce the issue @kelson42 was experiencing, using our existing code a picture saved for me and was instantly available in `Google Photos` under my albums, though it was named poorly.
> 
> The existing code can be found at `KiwixWebView.SaveHandler`

I dont have an Android 10 enabled device yet, how should I solve this?

**Comment by kelson42** [comment]
@GearGit You probably can not. Or maybe with the emulator.

**Comment by macgills** [comment]
The emulator provided by google. I can't remember if Genymotion have a free tier anymore but they were a saviour for many years

**Comment by 4shutosh** [comment]
I would like to work on this issue

**Comment by kelson42** [comment]
@4shutosh What would be your strategy to solve it?

**Comment by 4shutosh** [comment]
> The existing code can be found at `KiwixWebView.SaveHandler`
As @macgills suggested here.
I will try on the same 
And plan to make a specific folder for media within subfolders for audio(If required) and photos.
I think there wont be a issue doing this as we already have storage access for our app.

**Comment by yashasvi-goel** [comment]
Is this issue still unresolved? I'll love to work on it.

**Comment by 4shutosh** [comment]
> Is this issue still unresolved? I'll love to work on it.

Yes I am working on it!

**Comment by abdulwd** [comment]
I think we should use MediaStore. @macgills What do you suggest?

**Comment by macgills** [comment]
Whatever works? As I said I couldn't repro Kelson's issue.

I hate Mediastore apis but if it works go for it. If anything else works, go for it.

Check [the docs](https://developer.android.com/training/data-storage) they don't suck as much as they used to.

**Comment by 4shutosh** [comment]
Hey! @macgills Android 10 came with [Scoped Storage](https://developer.android.com/training/data-storage#scoped-storage) so the Standard File APIs won't work here.

The latest **Android 11** beta release also confirms [this]
(https://developer.android.com/preview/privacy/storage#scoped-storage). 

For Android 10 we still got  a `requestLegacyExternalStorage` for making a file directory (storage/emulated/0/Kiwix), but won't work in the future versions of Android.

Here is a [guide](https://proandroiddev.com/working-with-scoped-storage-8a7e7cafea3) by Fernando working with Scoped Storage using **MediaStore**

**Comment by macgills** [comment]
That is if we target it, we are not in a position to use Scoped Storage right now, we will request legacy storage when we upgrade our target sdk.

[Commonsware](https://commonsware.com/blog/2019/06/07/death-external-storage-end-saga.html) is the guy you go to for how utterly catastrophic the scoped storage change is

**Comment by 4shutosh** [comment]
Yes the targetSdk being less than 29 we can still use the older methods.
So making a easily accessible media directory on the ExternalStorage (outside the package folder) is still doable ignoring the Android 10 users, which was infact the reason behind this issue.

We can still make the directory `storage/emulated/0/Kiwix` for saving Pictures.
If so may I work on this?

**Comment by kelson42** [comment]
> We can still make the directory `storage/emulated/0/Kiwix`

@4shutosh This is not what I requested.

**Comment by 4shutosh** [comment]
Currently, on all devices the downloaded media is stored in `Android/media/com.kiwix.kiwix/`

> I expect to have it in "Downloads" or "Pictures"  like if I would take a picture with the camera or download a file with the browser.

@kelson Will it okay if we make it save in Downloads directory?

**Comment by macgills** [comment]
@4shutosh see the description
>If I "Save a media" (a picture here), it seems to put it in a place which is not easily available. I expect to have it in "Downloads" or "Pictures" like if I would take a picture with the camera or download a file with the browser. I have an Android 10.

After you save media with kiwix we want it to show up in other media viewer apps without the user having to do anything about it

**Comment by 4shutosh** [comment]
So it refers to `Environement.DIRECTORY_DOWNLOADS` or `Environement.DIRECTORY_DCIM` (For Camera ) or for Pictures `DIRECTORY_PICTURES` right?
I think it requires a workaround in KiwixWebView.handleMessage.
I will try implementing this.

**Comment by macgills** [comment]
Good luck!

**Comment by gouri-panda** [comment]
Is this issue still open @macgills? Can I work on this issue?

**Comment by macgills** [comment]
I would say this issue is still open, feel free to work on it @gouri-panda

**Comment by abhinav78910** [comment]
Hi @kelson42, can i work on this issue?

**Comment by abdulwd** [comment]
@gouri-panda good to assign this to @abhinav78910 or are you still working on this one?

**Comment by gouri-panda** [comment]
No, @abdulwd I'm not currently working this one.@abhinav78910 You can go ahead!

**Comment by Rishavgupta12345** [comment]
can i work on this issue

**Comment by 4shutosh** [comment]
There is a PR linked to this issue, take a look at what's going on there and ask the same on that thread.
